---
title: "QC Analysis and Integration 3rd and 4th DataSet"
author: "Camila Arcuschin; Ignacio Schor"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: true
    theme: cerulean
    code_folding: hide
    bibliography: references.bib
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE, dpi=500, cache=F, dev = c('png', 'pdf')) #dev = c('pdf', 'png') # If I want to export plots but don't look to the html (because it is too heavy). Remember to also add: clean = FALSE, as a parameter of the rmarkdown::render() function, when executing 
knitr::opts_knit$set(root.dir = "/data_husihuilke/shared/SMB/carcu/Parkinsons_project/")
```

```{r importpckg, include=FALSE}
library(Seurat)
library(tidyverse)
library(hdf5r)
library(pryr)
library(gridExtra)
library(grid)
library(ragg) # scaling plots
library(reticulate)
library(patchwork)
library(cowplot)
library(future)
library(htmltools)
library(DT)
library(ComplexHeatmap)
library(colorRamp2)
library(MAST)
plan(multisession, workers = 6)
#plan("multiprocess", workers = 6)
options(future.globals.maxSize = 4000 * 1024^2)
```

There are some warnings. Check them later

```{r Functions, include =FALSE}
#### Functions ####

# Create Seurat Object and filter by UMIs.p.cell and ribosomal percentage
Seu_filter<- function(dataX, UMIs.p.cell_lth, UMIs.p.cell_hth, percent.rb_th, treatment){
  # Implement some filter parameters to avoid working with a very big matrix
  # Gene must be expressed in at least 30 cell (of the 2452867 "cells")
  # Explore the data, rename the cells metadata
  GetAssayData(dataX)
  dataX[[]] %>% str()
  dataX@meta.data <- rename(dataX@meta.data, "ActiveGenes.p.cell" = "nFeature_RNA")
  dataX@meta.data <- mutate(dataX@meta.data, "UMIs.p.cell" = nCount_RNA)
  #Add MetaData
  is.ribo<-lapply(c("Rpl", "Rps"), function(x) grepl(x, rownames(dataX))) %>% unlist()
  ribo_genes<-lapply(c("Rpl", "Rps"), function(x) grep(x, rownames(dataX), value = T)) %>% unlist()
  ribo_umis <- subset(dataX, features = ribo_genes) %>% GetAssayData() %>% colSums()
  dataX[["percent.rb"]] <- 100*ribo_umis/dataX$UMIs.p.cell
  dataX[["percent.mt"]] <- PercentageFeatureSet(dataX, pattern = "Mt")
  # What would happen if I only kept the cell with more than 2000 UMIs (aprox remove the 7% of the distribution)
  dataX <- subset(dataX, subset = UMIs.p.cell > UMIs.p.cell_lth & UMIs.p.cell < UMIs.p.cell_hth & percent.rb < percent.rb_th)
  return(dataX)
}

get_legend<-function(myggplot){
  tmp <- ggplot_gtable(ggplot_build(myggplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}

# Log Normalize Seurat and select the 2000 most variable genes
norm_seurat<- function(all_samples) {
  all_samples <- NormalizeData(all_samples, normalization.method = "LogNormalize", scale.factor = 10000)
  all_samples <- FindVariableFeatures(all_samples, selection.method = "vst", nfeatures = 2000, loess.span = 0.6)
}

# Run PCA, knn and UMAP
UMAP_maker<-function(obj, dims, k.param){
  ScaleData(object = obj) %>%
    RunPCA(npcs = dims) %>%
    FindNeighbors(dims = 1:dims, k.param = k.param) %>% #Esta linea no har?a falta, igual creo que no cambiar?a nada porque siempre lo estoy pisando.
    RunUMAP(dims = 1:dims)
} # The reason why I have FindNeighbours here is because, if I want to replot UMAP whithout changing the cluster already assigned, I could do it

preprocess_RPCA <- function(obj.list, features) {
  obj.list<-ScaleData(object = obj.list, features = features) %>%
    RunPCA(features = features)
}

fisher<-function(x){
  matriz<-matrix(c(x[1], x[2], x[3], x[4]), nrow = 2)
  f_test<-fisher.test(matriz, alternative = "greater")
  data_fisher<-c(enrich = f_test[["estimate"]][["odds ratio"]], p_val = f_test[["p.value"]], conf_int = f_test[["conf.int"]])
  return(data_fisher)
}

create_random <- function(To, p_E, p_S, p_N){
  vec <- sample(c("EGFP", "SNCA", "EGFP_SNCA", "Neith"), To,  replace = T, prob = c(p_E-p_E*p_S, p_S-p_E*p_S, p_E*p_S, p_N))
  table(vec)
}

EGFP_SNCA <- function(df){
  nrow(filter(df, EGFP == "1" & hSNCA == "1" ))
}

EGFP <- function(df){
  nrow(filter(df, EGFP == "1" & hSNCA == "0" ))
}

SNCA <- function(df){
  nrow(filter(df, EGFP == "0" & hSNCA == "1" ))
}

Neith <- function(df){
  nrow(filter(df, EGFP == "0" & hSNCA == "0" ))
}

label2analyze <- function(FishTest){
  vect <- c()
  for (i in 1:length(clusters2analyze)){
    if (i != 3){
      vect[i] <- paste0("Cluster ", clusters2analyze[i], " = ", round(FishTest["enrich", clusters2analyze[i]], 2), " (p-adj = ", format(FishTest["FDR", clusters2analyze[i]], digits=3), ") ; ")
    } else {
      vect[i] <- paste0("Cluster ", clusters2analyze[i], " = ", round(FishTest["enrich", clusters2analyze[i]], 2), " (p-adj = ", format(FishTest["FDR", clusters2analyze[i]], digits=3), ") ;\n")
    }
  }
  return(vect)
}

explore_conditional_or <- function(x){
  matriz<-array(c(x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8]), dim = c(2,2,2))
  return(apply(matriz, 3, function(x) (x[1,1]*x[2,2])/(x[1,2]*x[2,1])))
}

woolf <- function(x) {
  x <- x + 1 / 2
  k <- dim(x)[3]
  or <- apply(x, 3, function(x) (x[1,1]*x[2,2])/(x[1,2]*x[2,1]))
  w <-  apply(x, 3, function(x) 1 / sum(1 / x))
  1 - pchisq(sum(w * (log(or) - weighted.mean(log(or), w)) ^ 2), k - 1)
}

woolf_test <- function(x){
  matriz<-array(c(x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8]), dim = c(2,2,2))
  return(woolf(matriz))
}

mantel_haenszel<-function(x){
  matriz<-array(c(x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8]), dim = c(2,2,2))
  f_test<-mantelhaen.test(matriz, alternative = "greater")
  data_mantel_haenszel<-c(enrich = f_test[["estimate"]][["common odds ratio"]], p_val = f_test[["p.value"]], conf_int = f_test[["conf.int"]])
  return(data_mantel_haenszel)
}

# Plotting
my_theme <- theme_classic() +
  theme(plot.title = element_text(size = 25, face = "bold"),
        axis.text.x = element_text(colour = "black", size = 14),
        axis.text.y = element_text(colour = "black", size = 14),
        legend.title = element_text(size = 15),
        legend.text = element_text(size = 15),
        axis.title.y = element_text(size = 15),
        axis.title.x = element_text(size = 15),
        strip.text.x = element_text(size = 15))

QC_plots <- function(SeuObj, extra_text, n_color) {
  Dist_QC <-
    VlnPlot(
      SeuObj,
      features = c(
        "ActiveGenes.p.cell",
        "UMIs.p.cell",
        "percent.mt",
        "percent.rb"
      ),
      ncol = 4,
      pt.size = 0,
      cols = n_color
    ) &
    theme(legend.position = "none",
          axis.title = element_text(size = 6),
          axis.text = element_text(size = 5),
          plot.title = element_text(size = 8, face = "bold"))
  stth <-     theme(
    legend.position = "none",
    axis.title = element_text(size = 6),
    axis.text = element_text(size = 5),
    plot.title = element_text(size = 8, face = "bold")
  )
  plot1 <-
    FeatureScatter(
      SeuObj,
      feature1 = "UMIs.p.cell",
      feature2 = "percent.mt",
      pt.size = 0.5,
      plot.cor =  F,
      cols = n_color,
      cells = sample(Cells(SeuObj))
    ) + stth
  plot2 <-
    FeatureScatter(
      SeuObj,
      feature1 = "UMIs.p.cell",
      feature2 = "ActiveGenes.p.cell",
      pt.size = 0.5,
      plot.cor =  F,
      cols = n_color,
      cells = sample(Cells(SeuObj))
    ) + stth
  plot3 <-
    FeatureScatter(
      SeuObj,
      feature1 = "ActiveGenes.p.cell",
      feature2 = "percent.rb",
      pt.size = 0.5,
      plot.cor =  F,
      cols = n_color,
      cells = sample(Cells(SeuObj))
    )+ stth
  plot4 <-
    FeatureScatter(
      SeuObj,
      feature1 = "UMIs.p.cell",
      feature2 = "percent.rb",
      pt.size = 0.5,
      plot.cor =  F,
      cols = n_color,
      cells = sample(Cells(SeuObj))
    ) + stth
  Scat_QC <-
    plot_grid(plot1, plot2, plot3, plot4, nrow = 1)
  QC_plot <- (Dist_QC/Scat_QC) +
    plot_annotation(
      title = paste0(
        extra_text,
        "Quality Control Metrics for > ",
        UMIs.p.cell_lowthreshold,
        " <",
        UMIs.p.cell_highthreshold,
        " UMI < ",
        percent.rb_threshold,
        " per.rb Matrix (",
        dim(SeuObj)[2],
        " Cells)"
      ),
      subtitle = "each number is a cluster",
      theme = theme(
        axis.text = element_text(size = 5),
        axis.title = element_text(size = 6),
        plot.title = element_text(size = 6, face = "bold")
      )
    )
  return(QC_plot)
}

UMAP_by_clust <-
  function(SeuObj,
           ClusterBreaks,
           ClusterLabels,
           pt_size,
           title_text,
           subtitle_text,
           color_a,
           color_b,
           do_label,
           n_color) {
    (
      UMAPa <-
        DimPlot(
          SeuObj,
          reduction = "umap",
          label = do_label,
          pt.size = pt_size
        ) +
         scale_colour_manual(values = n_color, breaks = ClusterBreaks, labels = ClusterLabels) +
        labs(
          title = title_text,
          subtitle = subtitle_text
          # color = color_a
        ) +
        theme(
          axis.title = element_text(size = 6),
          legend.title = element_text(size = 6),
          legend.text = element_text(size = 6),
          legend.position = "bottom",
          plot.title = element_text(size = 10, face = "bold"), axis.text = element_text(size = 0), axis.ticks = element_line(size = 0)
        ) +
        guides(color = guide_legend(
          title.position="top",
          override.aes = list(size = 1), nrow = 3
        ))
    )
    (
      UMAPb <-
        DimPlot(
          SeuObj,
          reduction = "umap",
          label = F,
          pt.size = pt_size,
          group.by = "Treat",
          cells = sample(Cells(SeuObj))
        ) +
        labs(
          title = "",
          subtitle = "",
          color = color_b
        ) +
        theme(
          legend.position = "right",
          axis.title = element_text(size = 6),
          legend.title = element_text(size = 6),
          legend.text = element_text(size = 6),
          plot.title = element_text(size = 10, face = "bold"), axis.text = element_text(size = 0), axis.ticks = element_line(size = 0)
        ) +
        guides(color = guide_legend(
          override.aes = list(size = 1), nrow = 2
        ))
    )
    return(list("UMAPa" = UMAPa, "UMAPb" = UMAPb))
  }


insert_minor <- function(major_labs, n_minor) {labs <- 
  c( sapply( major_labs, function(x) c(x, rep("", 4) ) ) )
labs[1:(length(labs)-n_minor)]}

UMAPs_for_grid <- function(SeuObj, subtitle_text){
  Idents(SeuObj) <- "Genes"
  UMAP4 <- DimPlot(SeuObj, reduction = "umap", order = T, label = F, pt.size = 1, label.size = 8, ncol = 1, cols = c("grey","#40a1e4", "#e48340", "#4c0e97")) +
    labs(title = "", subtitle = "EGFP-KASH+ hSNCA-A53T+ Cells") + theme(legend.position = "top", axis.text = element_text(size = 0), axis.ticks = element_line(size = 0)) + 
    guides(color = guide_legend(override.aes = list(size=4), nrow=1))
  legend <- get_legend(UMAP4)
  UMAP4 <- UMAP4 + theme(legend.position = "none")
  Idents(SeuObj) <- "seurat_clusters"
  UMAPa<-DimPlot(SeuObj, reduction = "umap", pt.size = 0.7, label = F, cols = colors_n) + theme(legend.position = "none", axis.text = element_text(size = 0), axis.ticks = element_line(size = 0))
  UMAPa_edit <- LabelClusters(UMAPa, id= "ident", size=4, repel = F) + labs(title = "", subtitle = subtitle_text) +
    theme(title = element_text(size = 15))
  return(list("UMAP4_split" = UMAP4, "UMAPa_edit" = UMAPa_edit, "UMAP4_legend" = legend))
}

FeaturePlotMultiple2<- function(obj, feature, color, ord, alf, minlim){
  # # the minimal and maximal of the value to make the legend scale the same. 
  # minimal<- min(obj[['RNA']]@data[feature, ])
  # maximal<- max(obj[['RNA']]@data[feature, ])
  ps<- vector("list", length(feature))
  for (i in 1:length(feature)) {
    if (ord){
      p<- FeaturePlot(obj, features = feature[i], min.cutoff = minlim, max.cutoff = "q90", pt.size = 0.00001, cols = c(alpha("lightgrey", alf), alpha(color[i], alf)), order = T) 
    } else {
      p<- FeaturePlot(obj, features = feature[i], min.cutoff = minlim, max.cutoff = "q90", pt.size = 0.00001, cols = c(alpha("lightgrey", alf), alpha(color[i], alf)), cells = sample(Cells(obj))) 
    }
    #ggtitle(feature[i]) + theme_void() +
    p <- p + theme(text = element_text(size = 8), axis.text = element_text(size = 5), plot.title = element_text(size = 10, face = "bold"), legend.position = "none")
    plot_name <- feature[i]
    ps[[i]] <- p
    names(ps)[[i]] <- plot_name
  }
  return(ps)
}

exp_cells<- function(x){
  sum(x>0)
}

TryDifferentPCs <- function(SeuObj, colors_n, dims){
UMAP.SeuObj <- DimPlot(SeuObj, group.by = "orig.ident", cells = sample(Cells(SeuObj))) + labs( subtitle = "") + theme_void(base_size = 10) 
UMAP.SeuObja <- DimPlot(subset(SeuObj, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend() 
UMAP.SeuObjb <- DimPlot(subset(SeuObj, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend() 
print(plot_grid(UMAP.SeuObj, NULL, UMAP.SeuObja, UMAP.SeuObjb, ncol = 2, nrow = 2))

plotl <- list()
for (i in 1:length(dims)){
SeuObj2 <- UMAP_maker(SeuObj, dims[i], 10)

UMAP.SeuObj2 <- DimPlot(SeuObj2, group.by = "orig.ident", cells = sample(Cells(SeuObj2))) + labs( subtitle = "") + theme_void(base_size = 10) 
colors_n <- rainbow(length(unique(SeuObj2@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuObj2@meta.data$seurat_clusters)
UMAP.SeuObj2a <- DimPlot(subset(SeuObj2, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend() 
UMAP.SeuObj2b <- DimPlot(subset(SeuObj2, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend()
title <- ggdraw() + 
  draw_label(paste0("dim = ", dims[i]))
plotl[[i]] <- plot_grid(title,
plot_grid(UMAP.SeuObj2, NULL, UMAP.SeuObj2a, UMAP.SeuObj2b, ncol = 2, nrow = 2),
ncol = 1, rel_heights = c(0.1, 1))
}
return(plotl)
}


MarkerPlot <- function(SeuObj){
# Neuronal markers:
Nordenergic <- c("Th", "Dbh") #"Th": Dopamine synthesis enzyme 
NoradrenalineTransporter <- "Slc6a2"
GABAergic <- c("Gad1", "Gad2") #"Gad1", "Gad2" : GABA synthesis enzymes
Glutamatergic <- c("Slc17a7", "Slc17a6", "Slc17a8") #"Slc17a7", "Slc17a6", "Slc17a8": Glutamate transporters
Cholinergic <- c("Slc18a3", "Slc5a7") #"Slc18a3", "Slc5a7": Acetylcholine transporters "Gm5741"?
DopamineTransporter <- "Slc6a3" #"Slc6a3": Dopamine Transporter
AmineTranporter <- c("Slc18a2", "Vat1") #"Slc18a2": Monoamine transporter (dopamine, norepinephrine, serotonin, and histamine) #"Vat1": Vesicle amine transporter
Serotoninergic <- c("Tph1", "Htr1a", "Htr1b", "Htr1f", "Htr1d",  "Htr2c") #"Tph1": Serotonin synthesis enzyme  #"Htr1a", "Htr1b", "Htr1f", "Htr1d",  "Htr2c": Serotonin transporter
AmineDegradetion <- c("Comt", "Maoa", "Maob") #"Comt", "Maoa", "Maob": Amine degradetion

# Plot neuronal markers on X DS
DefaultAssay(SeuObj) <- "RNA" # Th gene is not a variable feature in the INTE assay
th <- theme(text = element_text(size = 8), plot.title = element_text(size = 10, face = "bold"), axis.text = element_text(size = 0), axis.ticks = element_line(size = 0))
plot1 <- FeaturePlot(SeuObj, features = c(Nordenergic, NoradrenalineTransporter, "EGFP-KASH"), pt.size = 0.00001, order =  T, ncol = length(c(Nordenergic, NoradrenalineTransporter, "EGFP-KASH"))) & th #normalized counts
SeuObj@meta.data$Thg <- ifelse(GetAssayData(SeuObj, slot= "counts")["Th",]>0, 1, 0) %>% as.factor()
SeuObj@meta.data$Dbhg <- ifelse(GetAssayData(SeuObj, slot= "counts")["Dbh",]>0, 1, 0) %>% as.factor()
SeuObj@meta.data$Slc6a2g<- ifelse(GetAssayData(SeuObj, slot= "counts")["Slc6a2",]>0, 1, 0) %>% as.factor()
plot2 <- DimPlot(SeuObj, group.by = "Thg", order = T) + th
plot3 <- DimPlot(SeuObj, group.by = "Dbhg", order = T) + th
plot4 <- DimPlot(SeuObj, group.by = "Slc6a2g", order = T) +th
layout1<-"
ABC
DFG
HIJ
KLM
"
plot5<- FeaturePlotMultiple2(SeuObj, feature = c(GABAergic, Glutamatergic, Cholinergic, DopamineTransporter, AmineTranporter), color = rep(c("deepskyblue3", "limegreen", "chocolate2", "violetred2", "red"), times = c(2,3,2,1,2)), ord = T, alf = 0.5, minlim = "q1")
plot5<- wrap_plots(plot5, guides = "keep", design = layout1) & th # Slc18a3, Slc17a7, Slc17a8 very low expression
plot6<- FeaturePlotMultiple2(SeuObj, feature = c(Serotoninergic, AmineDegradetion), color = rep(c("#6e2c00", "darkorange"), times = c(6,3)), ord = T, alf = 0.5, minlim = "q1")
plot6<- wrap_plots(plot6, guides = "keep", design = layout1) & th # Tph1, Maob very low expression
palfa <- plot_grid(plot1 , plot_grid(plot2, plot3, plot4, NULL, nrow = 1), ncol = 1)
pbeta <- plot_grid(plot5, plot_grid(plot6, NULL, ncol = 1, rel_heights = c(5, 1)), nrow = 1)
print(plot_grid(palfa, pbeta, ncol = 1, rel_heights = c(0.5, 1)))
}

DotPlotRemake <- function(SeuObj, markers){
markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) %>%
    slice_head(n = 2) %>%
    ungroup() -> top2
dp <- DotPlot(SeuObj, features = unique(top2$gene), cluster.idents = T) + RotatedAxis()
id.levels <- levels(dp$data$id)
id.levels<- id.levels[(id.levels %in% top2$cluster)]
id.levels.df <- data.frame(cluster = id.levels, ord = 1:length(id.levels))
top2 <- left_join(id.levels.df, top2)
ptop2 <- DotPlot(SeuObj, features = unique(top2$gene), cluster.idents = T, dot.scale = 3) + RotatedAxis() + theme(axis.text = element_text(size = 7))
print(ptop2)
markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) %>%
    slice_head(n = 3) %>%
    ungroup() -> top3
dp <- DotPlot(SeuObj, features = unique(top3$gene), cluster.idents = T) + RotatedAxis()
id.levels <- levels(dp$data$id)
id.levels<- id.levels[(id.levels %in% top3$cluster)]
id.levels.df <- data.frame(cluster = id.levels, ord = 1:length(id.levels))
top3 <- left_join(id.levels.df, top3)
ptop3 <- DotPlot(SeuObj, features = unique(top3$gene), cluster.idents = T, dot.scale = 3) + RotatedAxis() + theme(axis.text = element_text(size = 7))
print(ptop3)
return(list("top2" = top2, "top3" = top3))
}


DotPlotRemake2 <- function(SeuObj, markers){
markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) %>%
    slice_head(n = 2) %>%
    ungroup() -> top2
dp <- DotPlot(SeuObj, features = unique(top2$gene), cluster.idents = T) + RotatedAxis()
id.levels <- levels(dp$data$id)
id.levels<- id.levels[(id.levels %in% top2$cluster)]
id.levels.df <- data.frame(cluster = id.levels, ord = 1:length(id.levels))
top2 <- left_join(id.levels.df, top2)
ptop2 <- DotPlot(SeuObj, features = unique(top2$gene), cluster.idents = T, dot.scale = 3) + RotatedAxis() + theme(axis.text = element_text(size = 7))
markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) %>%
    slice_head(n = 3) %>%
    ungroup() -> top3
dp <- DotPlot(SeuObj, features = unique(top3$gene), cluster.idents = T) + RotatedAxis()
id.levels <- levels(dp$data$id)
id.levels<- id.levels[(id.levels %in% top3$cluster)]
id.levels.df <- data.frame(cluster = id.levels, ord = 1:length(id.levels))
top3 <- left_join(id.levels.df, top3)
ptop3 <- DotPlot(SeuObj, features = unique(top3$gene), cluster.idents = T, dot.scale = 3) + RotatedAxis() + theme(axis.text = element_text(size = 7))
print(ptop3)
return(list("top2" = top2, "top3" = top3))
}

DoHeatmapRemake <- function(SeuObj, assesed_genes, lims, colors_n, which_type, gene_names){

my_palette <- colorRampPalette(c("magenta", "#440154ff", "black", "#fde725ff"), space = "Lab")(n = 200)

SeuObj <- ScaleData(SeuObj, features = assesed_genes, assay = "RNA", verbose = F)
assesed_genes <- GetAssayData(SeuObj, slot = "scale.data") %>% rownames()
scale_data <- GetAssayData(SeuObj, slot = "scale.data")

ha = HeatmapAnnotation(Clusters = SeuObj$seurat_clusters, DS = SeuObj$orig.ident, col = list(Clusters = colors_n, DS = c("`3rd`" = "#00BFC4", "`4th`" = "#F8766D")))
clust_fact <- SeuObj$seurat_clusters %>% as.character()
names(clust_fact) <- NULL

set.seed(0)

scale_data[scale_data >= lims] <- lims
scale_data[scale_data <= -lims] <- -lims

gene_names <- rep(gene_names, nrow(scale_data))
row.labels <- ifelse(gene_names, rownames(scale_data), rep("", nrow(scale_data)))

HM_OD <- Heatmap(scale_data, name = "Gene Scaled Exp",
                 col = colorRamp2(seq(-(lims-0.5),(lims-0.5),0.01*(lims-0.5))[-1], my_palette),
                 cluster_columns = cluster_within_group(scale_data, clust_fact),
                 top_annotation = ha, show_row_dend = T, show_column_dend = F,
                 cluster_rows = T,
                 row_labels = row.labels,
                 row_names_gp = gpar(fontsize = 5),
                 column_labels = rep("", ncol(scale_data)),
                 cluster_column_slices = T,
                 clustering_distance_columns =  "euclidean",
                 column_title_rot = 90, row_title = which_type,
                 row_title_gp = gpar(col = "black", fontsize = 10),
                 row_title_rot = 90,
                 column_km_repeats = 100,
                 row_km_repeats = 100)
        
HM_OD <- draw(HM_OD)

}


DoHeatmapRemakeAsSeurat <- function(SeuObj, markers, assesed_genes, lims, colors_n, which_type, gene_names){

my_palette <- colorRampPalette(c("magenta", "#440154ff", "black", "#fde725ff"), space = "Lab")(n = 200)

SeuObj <- ScaleData(SeuObj, features = assesed_genes, assay = "RNA", verbose = F)
scale_data <- GetAssayData(SeuObj, slot = "scale.data")

seurat_clusters <- SeuObj$seurat_clusters
cells <- Cells(SeuObj)
df <- data.frame(seurat_clusters, cells)
cluster_order <- markers$cluster %>% unique() %>% as.character()
df <- left_join(data.frame(seurat_clusters = cluster_order), df) %>% na.omit()
head(df)

ha = HeatmapAnnotation(Clusters = SeuObj$seurat_clusters[df$cells], DS = SeuObj$orig.ident[df$cells], col = list(Clusters = colors_n, DS = c("`3rd`" = "#00BFC4", "`4th`" = "#F8766D")))
clust_fact <- SeuObj$seurat_clusters[df$cells] %>% as.character()
names(clust_fact) <- NULL

scale_data[scale_data >= lims] <- lims
scale_data[scale_data <= -lims] <- -lims

scale_data <- scale_data[assesed_genes, df$cells]
gene_names <- rep(gene_names, nrow(scale_data))
row.labels <- ifelse(gene_names, rownames(scale_data), rep("", nrow(scale_data)))

HM_OD <- Heatmap(scale_data, name = "Gene Scaled Exp",
                 col = colorRamp2(seq(-(lims-0.5),(lims-0.5),0.01*(lims-0.5))[-1], my_palette),
                 cluster_columns = F,
                 cluster_rows = F,
                 top_annotation = ha, show_row_dend = F, show_column_dend = F,
                 row_labels = row.labels,
                 row_names_gp = gpar(fontsize = 5),
                 column_labels = rep("", ncol(scale_data)),
                 cluster_column_slices = F,
                 column_title_rot = 90, row_title = which_type,
                 row_title_gp = gpar(col = "black", fontsize = 10),
                 row_title_rot =  90) 
        
HM_OD <- draw(HM_OD)
}


```

## Import Input

Import input from the 3rd and the 4th DataSet (DS). Each of the inputs are already integrated in their HTO expression. [Here](https://github.com/satijalab/seurat/discussions/5681) is another post in which someone else integrate many datasets already integrated to avoid time and RAM consuming computing. However, it is worth mentioning that since all this has been made with the same sequencing facility (10X) and that in the integration some genes are lost (only 2000 recovered), maybe it is also important to try to integrate the data all together (will the same genes be recovered? I guess that yes since all conditions must match in their "variable genes")

```{r ImportInput, include = FALSE}
#### Run ####

##  Data ####
## 3rd and 4th DataSet to integrate

# Pagina en donde otra persona integra datos pre-integrados
#https://github.com/satijalab/seurat/discussions/5681

## 4th DataSet to integrate
## filter parameters ####
# No filter, so:
date1 = 231122
date2 = 231128
File1 <- paste0("Outputs/", date2, "_", date1, "_SeuInt_4th_fil_int_mayt1000lowt60000_lowtInfrb_LC_SoupXOut_reAnalized.rds")
`4th` <- readRDS(File1)
DefaultAssay(`4th`) <- "RNA"
`4th`@meta.data$orig.ident <- "`4th`"


## 3rd DataSet to integrate
## filter parameters ####
# No filter, so:
date1 = 231122
date2 = 231128
File2 <- paste0("Outputs/", date2, "_", date1, "_SeuInt_3rd_fil_int_mayt1000lowt60000_lowtInfrb_LC_SoupXOut_reAnalized.rds")
`3rd` <- readRDS(File2)
DefaultAssay(`3rd`) <- "RNA"
`3rd`@meta.data$orig.ident <- "`3rd`"

date = 231129

UMIs.p.cell_lowthreshold = 1000
UMIs.p.cell_highthreshold = 60000
percent.rb_threshold = Inf

```


```{r QCwoInt, out.width= '100%', fig.asp=1}
## Others
clusters2analyze = c(0,1,2,3,4,5,6,7,8,9,11,12) %>% as.character() #ok 18 15 14 13
text4stackbar = c(0,1,2,3,4,5,9) %>% as.character() #ok
text4stackbarEGFP = c(0,1,2,7,9,11) %>% as.character() #ok 15


## Combine objects ####
SeuMer <- merge(`3rd`, `4th`, add.cell.ids = c("`3rd`", "`4th`"), merge.data = TRUE)
SeuMer@meta.data$orig.ident <- factor(SeuMer@meta.data$orig.ident, levels = c("`3rd`", "`4th`"))
```

## Plot the UMAP with the 2 DS non integrated with each other

```{r UMAPwoInt}
## Plot data without integration
DefaultAssay(SeuMer) <- "RNA"
SeuMer <- FindVariableFeatures(SeuMer) #reduction = "rpca" The following arguments are not used: reduction (date = 230523)
SeuMer<-UMAP_maker(SeuMer, dims = 50, k.param = 10)
UMAP.SeuMer1 <- DimPlot(SeuMer, group.by = "orig.ident", reduction = "umap", pt.size = 0.1, cells = sample(Cells(SeuMer))) + labs( subtitle = "") + theme_void(base_size = 10) 
colors_n <- rainbow(length(unique(SeuMer@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuMer@meta.data$seurat_clusters)
UMAP.SeuMer2a <- DimPlot(subset(SeuMer, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend()
UMAP.SeuMer2b <- DimPlot(subset(SeuMer, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend() 

# Obtain the ASW (Average Silhouette Width) batch score (integration quality parameter)
#ElbowPlot(SeuMer, ndims = 50)
obj_dims = 30 # Dimensions to use
d <- stats::dist(Embeddings(SeuMer, reduction = "pca")[,1:obj_dims], method = "euclidean") # Create Matrix of Euclidean distances between all pair cells
batches <- tibble(CellNames = dimnames(SeuMer)[[2]], Batch = SeuMer@meta.data$orig.ident) # Set the Batch correspondence to each cell
batches <- mutate(batches, Batch_int = as.numeric(as.factor(Batch)))
sil_batch <- cluster::silhouette(x = batches$Batch_int - 1, dist = d) # Calculate SW for batch effect for each cell
rm(batches, d)
ASW_batch <- mean(tibble::as_tibble(sil_batch[, ])$sil_width) # Calculate the Average Silhoutte Width for batch effect

text <- ggplot(data.frame(x=1, y=1), aes(x, y)) + geom_text(label = paste0("The ASW batch value is ", round(ASW_batch, 4))) + theme_void(base_size = 6)
plot_grid(UMAP.SeuMer1, text, UMAP.SeuMer2a, UMAP.SeuMer2b, ncol = 2, nrow = 2)
```

## Integrate the 2 DS with each other

The way to do the integration is the following:

-   Select the features to integrate: this are the joint 2000 top variable features in both DS. The variable features are already computed separately on each DS in past scripts and here these features are positively ranked for being variable in both DS.

-   Find the pair of cells between DSs that can match the best in the reduced PCA space (anchors) of each DS, taking the 3rd DS PCA as a reference (in this case having only 2 DS, the reference doesn't matter). It is like finding the nearest neighborhood of cell in its neighborhood PCA projection

-   Integrate both DS subtracting the expression values of the selected features from one member of the anchor to the other member. Without changing the values of the cells that cannot be anchored?. A weight is also applied to penalize the distance between each anchor. In my particular case, I integrated using the selected features plus some neuronal, glial, and capilar cell-markers of interest, plus the transgenes.

-   Values for k.anchor, k.filter, k.score, max.features and k.weight are obtained by the ones tested in the @otero-garcia2022 paper DS (from Inma Cobos lab) which retrieved best ASW batch scores (see Alzheimer_project/Scripts/scRNAseq_integration_RCPA_labmeet.html to see the values)

```{r Integrate}
## Integrate Samples ####
# Prepare the objects to integrate
DefaultAssay(`3rd`) <- "integrated"
DefaultAssay(`4th`) <- "integrated"
SeuObj<- list(`3rd`, `4th`) # Create a list of Seurat Objects to integrate
#SeuObj <- SplitObject(SeuMer, split.by = "orig.ident")
SeuIntFeatures <- SelectIntegrationFeatures(SeuObj) # Features (genes) to use in the integration
#lapply(SeuObj, ElbowPlot, ndims = 50)
obj_dims = 30 # Dimensions to use
rm(SeuMer, `3rd`, `4th`)

# Cell-cell pairings (anchors) to do the integration
# Method: Reciprocal PCA
anchors <- FindIntegrationAnchors(SeuObj, reduction = "rpca", reference = 1,  dims = 1:obj_dims, anchor.features = SeuIntFeatures, scale = F, k.anchor = 10, k.filter = 100,  k.score = 20, max.features = 100) # FindAnchors future.seed=TRUE
# Features to include in the integration (markers of different cell types, including Locus Ceruleus (LC))
LCmark <- c("Dbh", "Th", "Slc6a2","Chrna7", "Maoa", "Rlim", "Sec23a", "Ppp2r2d", "Usp34") # Removed markers:  "Adra1a" "Ovgp1", "Smad9", "Gal", "Galr1", "Npy", falta "Galr3" y "Galr2"
Neuron <- c("Rbfox3") # General neurons
Ex <- c("Slc17a6", "Slc17a7") # Excitatory neurons
Inh <- c("Gad1") # Inhibitory neuron
Astrocytes <- c("Gfap", "Gja1") # Removed markers: Blbp = Fabp7; 
Microglia <- c("C1qb") # Removed markers: Aif1 = Iba1
OPC <- c("Olig1", "Tnr")
Oligodendrocyte <- c("Tfrc", "Plp1", "Mbp") # Tfrc = Tfr
Fibro <- c("Dcn") # Fibroblast
Capil <- c("Flt1", "Rgs5")
markers <- c("EGFP-KASH", "hSNCA-A53T", "mCherry",  "Snca")
# GliaMark<-c("Aldh1l1", "Aldoc", "Olig2", "Cnp")

# Integrate data including in the final matrix some marker genes
SeuInt <- IntegrateData(anchorset = anchors, dims = 1:obj_dims, k.weight = 20, new.assay.name = "INTE", features.to.integrate = c(SeuIntFeatures, LCmark, Neuron, Ex, Inh, Fibro, Astrocytes, Microglia, Oligodendrocyte, OPC, Capil, markers))
# Porfa mirar: https://github.com/satijalab/seurat/issues/6341 para el error "x[[jj]][iseq] <- vjj :number of items to replace is not a multiple of replacement length"
rm(anchors, SeuObj)
SeuInt@meta.data$orig.ident <- factor(SeuInt@meta.data$orig.ident, levels = c("`3rd`", "`4th`"))
DefaultAssay(SeuInt) <- "INTE" # (changed the order of this on date = 230523, originally after UMAP_maker)
SeuInt<-UMAP_maker(SeuInt, dims = 50, k.param = 10)

# Obtain the ASW (Average Silhouette Width) batch score (integration quality parameter)
#ElbowPlot(SeuInt, ndims = 50)
obj_dims = 30 # Dimensions to use
d <- stats::dist(Embeddings(SeuInt, reduction = "pca")[,1:obj_dims], method = "euclidean") # Create Matrix of Euclidean distances between all pair cells
batches <- tibble(CellNames = dimnames(SeuInt)[[2]], Batch = SeuInt@meta.data$orig.ident) # Set the Batch correspondence to each cell
batches <- mutate(batches, Batch_int = as.numeric(as.factor(Batch)))
sil_batch <- cluster::silhouette(x = batches$Batch_int - 1, dist = d) # Calculate SW for batch effect for each cell
rm(batches, d)
ASW_batch <- mean(tibble::as_tibble(sil_batch[, ])$sil_width) # Calculate the Average Silhoutte Width for batch effect
```

## Plot the UMAP of the integrated DS

```{r UMAPintegratedDS}
# Find clusters and color UMAP by clusters with Louvain and a high number of iterations
UMAP.SeuInt1 <- DimPlot(SeuInt, group.by = "orig.ident", cells = sample(Cells(SeuInt))) + labs( subtitle = "") + theme_void(base_size = 10) 
colors_n <- rainbow(length(unique(SeuInt@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuInt@meta.data$seurat_clusters)
UMAP.SeuInt2a <- DimPlot(subset(SeuInt, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend() 
UMAP.SeuInt2b <- DimPlot(subset(SeuInt, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend()

text <- ggplot(data.frame(x=1, y=1), aes(x, y)) + geom_text(label = paste0("The ASW batch value is ", round(ASW_batch, 4))) + theme_void(base_size = 6)
plot_grid(UMAP.SeuInt1, text, UMAP.SeuInt2a, UMAP.SeuInt2b, ncol = 2, nrow = 2)
```

# Re-clusterize the integrated DS

```{r Re-clusterIntegratedDS, include=FALSE, results='hide'}

ElbowPlot(SeuInt, ndims = 50, reduction = "pca")

SeuInt_1 <- FindNeighbors(SeuInt, dims = 30, k.param = 10, compute.SNN = F)
SeuInt_1 <- FindClusters(SeuInt_1, resolution = 0.05, n.start = 100, n.iter = 10)
```

```{r}
UMAP.SeuInt_11 <- DimPlot(SeuInt_1, group.by = "orig.ident", cells = sample(Cells(SeuInt_1))) + labs( subtitle = "") + theme_void(base_size = 10) 
colors_n <- rainbow(length(unique(SeuInt_1@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuInt_1@meta.data$seurat_clusters)
UMAP.SeuInt_12a <- DimPlot(subset(SeuInt_1, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend() 
UMAP.SeuInt_12b <- DimPlot(subset(SeuInt_1, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend() 
plot_grid(UMAP.SeuInt_11, NULL, UMAP.SeuInt_12a, UMAP.SeuInt_12b, ncol = 2, nrow = 2)
```

## QC plots of the integrated DS

```{r QCintegratedDS, out.width= '100%', fig.asp= 0.8}
# Plot QC graphics per cluster ####
QC_plots(SeuInt_1, "", colors_n)
```

A high percentage of ribosomal UMIs do imply that those cells are cell at division or with higher metabolism?

[Ayshwarya Subramanian et al.,](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9793662/) says that "mitochondrial transcript abundance is dependent on cellular physiology [13], and metabolically active tissues (e.g., muscle, kidney) have higher mitochondrial transcript content [14, 15]. Ribosomal protein gene expression has also been shown to vary by tissue [16] in human adults and mice [17]. Although biological variability in ribosomal protein gene expression has been reported [18], ribosomal protein gene expression is often conflated with technical artifacts or housekeeping transcription activity during analysis."

```{r, out.width= '120%', fig.asp= 0.6}
QC_int1 <- VlnPlot(SeuInt_1, features  = c("ActiveGenes.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int2 <- VlnPlot(SeuInt_1, features  = c("UMIs.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.text = element_text(size = 4), legend.key.size = unit(0.01, "npc"), plot.margin = unit(rep(0, 4), "npc"))
legend <- get_legend(QC_int2)
QC_int2 <- QC_int2 + theme(legend.position = "none")
QC_int3 <- VlnPlot(SeuInt_1, features  = c("percent.mt"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int4 <- VlnPlot(SeuInt_1, features  = c("percent.rb"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
plot_grid(QC_int1, QC_int2, legend, QC_int3, QC_int4, nrow = 2, ncol = 3, rel_widths = c(1.2,1.2, 0.1), scale = 1)
```

```{r}
SeuInt_1 <- subset(SeuInt_1, subset = seurat_clusters != "16" & seurat_clusters != "17" & seurat_clusters != "20" & seurat_clusters != "21" &seurat_clusters != "22" & seurat_clusters != "23" & seurat_clusters != "24")
SeuInt_1$seurat_clusters <- factor(SeuInt_1$seurat_clusters, levels = unique(SeuInt_1$seurat_clusters))
Idents(SeuInt_1) <- SeuInt_1$seurat_clusters
colors_n <- rainbow(length(unique(SeuInt_1@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuInt_1@meta.data$seurat_clusters)
```

## TryDifferentPCs

```{r}
UMAPs <- TryDifferentPCs(SeuInt_1, dims = c(10, 15, 20), colors_n)
UMAPs[[1]]
UMAPs[[2]]
UMAPs[[3]]
```

And ploting the last QC metrics but fir every DS:

```{r, out.width= '120%', fig.asp= 0.6}
QC_int1 <- VlnPlot(SeuInt_1, features  = c("ActiveGenes.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int2 <- VlnPlot(SeuInt_1, features  = c("UMIs.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.text = element_text(size = 4), legend.key.size = unit(0.01, "npc"), plot.margin = unit(rep(0, 4), "npc"))
legend <- get_legend(QC_int2)
QC_int2 <- QC_int2 + theme(legend.position = "none")
QC_int3 <- VlnPlot(SeuInt_1, features  = c("percent.mt"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int4 <- VlnPlot(SeuInt_1, features  = c("percent.rb"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 4), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
plot_grid(QC_int1, QC_int2, legend, QC_int3, QC_int4, nrow = 2, ncol = 3, rel_widths = c(1.2,1.2, 0.1), scale = 1)
```


```{r QCintegratedDS_separateRep, out.width= '100%', fig.asp= 0.6}
QC_plots(subset(SeuInt_1, subset = orig.ident == "`3rd`"), "3rd DS", colors_n)

QC_plots(subset(SeuInt_1, subset = orig.ident == "`4th`"), "4th DS", colors_n)
```

## The normalized expression of neuronal types markers

-   Slc6a2 = Noradrenaline Transporter (NET)

-   Th = Catalize L-DOPA production (precursor of Dopamine)

-   Dbh = Catalize Noradrenaline production from Dopamine

-   lightblue = GABAergic

-   lightgreen = Glutamatergic

-   brown = Cholinergic

-   violet = DopamineTransporter

-   red = AmineTranporter

-   darkbrown = Serotoninergic

-   orange = AmineDegradetion

#### In Ctrl
```{r MarkerExpression, out.width= '150%', out.height= '50%', fig.asp=1.2}
SeuInt_1_Ctrl <- subset(SeuInt_1, subset = Treat == "Ctrl")
MarkerPlot(SeuInt_1_Ctrl)
```


## Make differential expression of neurons and then a dot plot to see if there are similar markers to the ones in Caramia 2023 paper.

```{r}
DefaultAssay(SeuInt_1_Ctrl) <- "RNA"
SeuInt_1_n_Ctrl <- subset(SeuInt_1_Ctrl, subset = seurat_clusters != "1" & seurat_clusters != "2" & seurat_clusters != "3" & seurat_clusters != "4")
SeuInt_1_n_Ctrl_markers <- FindAllMarkers(SeuInt_1_n_Ctrl, test.use = "MAST", latent.vars = c("orig.ident", "Age"), only.pos =  T)
tops_2and3 <- DotPlotRemake(SeuInt_1_n_Ctrl, SeuInt_1_n_Ctrl_markers)
```

Clusters seem to be well defined by the expression of their top DEGs, despite from cluster 0 and I have doubts about the definition of cluster 8 and 14. 

If I compare this with the data of Caramia 2023, I can see that all clusters resemble some already described cluster, by exception of:

-   cluster 5 (marked by Arpp21, "The encoded protein is enriched in the caudate nucleus and cerebellar cortex. A similar protein in mouse may be involved in regulating the effects of dopamine in the basal ganglia" and Tcf7l2, " plays a key role in the Wnt signaling pathway", **maybe a new cluster**)

-   cluster 13 (Vomeronasal genes and Gm32647) 

-   cluster 10 (marked by mt-Co1, mt-Atp6 and Zic1, development. Is this a product of the technique, resembling many mRNAs in the sequencing "soup"?)

-   cluster 6 (marked by Pbx1, Nfib and Satb1, neurodevelopmental genes)

-   cluster 14 (marked by Zfpm2 and 2600014E21Rik, first one developmental, second one lncRNA) 

-   cluster 9 (marked by Sema5a, axonogenesis, Adarb2, RNA editing, and Ebf2, development)

-   cluster 7 (marked by Dach1, which "cell fate determination during development" and Cdh8, "putatively involved in synaptic adhesion, axon outgrowth and guidance", probably representing **cells under development** and Tcf4)

-   cluster 8 (marked by Pde4d, a second menssager, slightly by Lmx1a, associated with ParkinsonÅ› Disease and important in development of dopamine producing neurons during embryogenesis and Chst9, a sulfotransferrase)

-   cluster 12 (marked by Zfp804b, TF enable to bind ion and act in the nucleus, Epha3, ERK signaling, Ghr, PI3K-Akt signaling) 

-   cluster 15 (marked by monaminergic and noradrenergic transports or noradrenaline biosinthetic enzymes)

-   cluster 19 (marked by EGFP-KASH, Gm41333 and 1600010M07Rik, two lncRNA, which probably means that this cluster **is a product of viral infection itself** and not something biologically meaninful), 

-   cluster 11 (marked by Nr4a2, hormone related to Parkinson Disease and essential for essential for development and mantainance of mdDA neurons, C1ql4, maybe involved in synapse or differentiation and Ebf1, TF of B-cell specification and commitment).

-   cluster 18 (marked by Nos1, nitric oxide synthase, Fgfr1, fibroblast growth factor, and Slc6a5, a choline transpoter, **maybe acetylcholinergic cells**).


# Re-clustering

```{r}
SeuInt_2 <- FindNeighbors(SeuInt, dims = 50, k.param = 10, compute.SNN = F)
SeuInt_2 <- FindClusters(SeuInt_2, resolution = 0.2, n.start = 100, n.iter = 10)
dim(SeuInt_2)
```

```{r}
UMAP.SeuInt_21 <- DimPlot(SeuInt_2, group.by = "orig.ident", cells = sample(Cells(SeuInt_2))) + labs( subtitle = "") + theme_void(base_size = 10) 
colors_n <- rainbow(length(unique(SeuInt_2@meta.data$seurat_clusters)))
names(colors_n) <- unique(SeuInt_2@meta.data$seurat_clusters)
UMAP.SeuInt_22a <- DimPlot(subset(SeuInt_2, subset = orig.ident == "`3rd`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "3rd") + theme_void(base_size = 10) + NoLegend() 
UMAP.SeuInt_22b <- DimPlot(subset(SeuInt_2, subset = orig.ident == "`4th`"), group.by = "seurat_clusters", order = T, label = T, cols = colors_n) + labs(title = "4th") + theme_void(base_size = 10) + NoLegend() 
plot_grid(UMAP.SeuInt_21, NULL, UMAP.SeuInt_22a, UMAP.SeuInt_22b, ncol = 2, nrow = 2)
```

## QC plots of the integrated DS

```{r, out.width= '100%', fig.asp= 0.8}
# Plot QC graphics per cluster ####
QC_plots(SeuInt_2, "", colors_n)
```

A high percentage of ribosomal UMIs do imply that those cells are cell at division or with higher metabolism?

[Ayshwarya Subramanian et al.,](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9793662/) says that "mitochondrial transcript abundance is dependent on cellular physiology [13], and metabolically active tissues (e.g., muscle, kidney) have higher mitochondrial transcript content [14, 15]. Ribosomal protein gene expression has also been shown to vary by tissue [16] in human adults and mice [17]. Although biological variability in ribosomal protein gene expression has been reported [18], ribosomal protein gene expression is often conflated with technical artifacts or housekeeping transcription activity during analysis."

## Get the comparison of past clusters with actual clusters:

```{r}
past_df <- data.frame(cells = Cells(SeuInt_1), past_clusters = SeuInt_1$seurat_clusters)
actual_df <- data.frame(cells = Cells(SeuInt_2), actual_clusters = SeuInt_2$seurat_clusters)
past_actual_df <- left_join(past_df, actual_df) %>% select("past_clusters", "actual_clusters")
past_actual_df %>% unique()
```

```{r, fig.width= 10, fig.asp= 0.4}
QC_int1 <- VlnPlot(SeuInt_2, features  = c("ActiveGenes.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int2 <- VlnPlot(SeuInt_2, features  = c("UMIs.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.text = element_text(size = 4), legend.key.size = unit(0.01, "npc"), plot.margin = unit(rep(0, 4), "npc"))
legend <- get_legend(QC_int2)
QC_int2 <- QC_int2 + theme(legend.position = "none")
QC_int3 <- VlnPlot(SeuInt_2, features  = c("percent.mt"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int4 <- VlnPlot(SeuInt_2, features  = c("percent.rb"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
plot_grid(QC_int1, QC_int2, legend, QC_int3, QC_int4, nrow = 2, ncol = 3, rel_widths = c(1.2,1.2, 0.1), scale = 1)
```

```{r}
SeuInt_2 <- subset(SeuInt_2, subset = seurat_clusters != "20" & seurat_clusters != "21" & seurat_clusters != "22" & seurat_clusters != "25" &seurat_clusters != "26" & seurat_clusters != "27" & seurat_clusters != "28" & seurat_clusters != "29" & seurat_clusters != "30")
SeuInt_2$seurat_clusters <- factor(SeuInt_2$seurat_clusters, levels = unique(SeuInt_2$seurat_clusters))
Idents(SeuInt_2) <- SeuInt_2$seurat_clusters
dim(SeuInt_2)
```

## TryDifferentPCs

```{r}
UMAPs <- TryDifferentPCs(SeuInt_2, dims = c(10, 15, 20), colors_n)
UMAPs[[1]]
UMAPs[[2]]
UMAPs[[3]]
```

And ploting the last QC metrics but fir every DS:

```{r, fig.width= 10, fig.asp= 0.4}
QC_int1 <- VlnPlot(SeuInt_2, features  = c("ActiveGenes.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int2 <- VlnPlot(SeuInt_2, features  = c("UMIs.p.cell"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.text = element_text(size = 4), legend.key.size = unit(0.01, "npc"), plot.margin = unit(rep(0, 4), "npc"))
legend <- get_legend(QC_int2)
QC_int2 <- QC_int2 + theme(legend.position = "none")
QC_int3 <- VlnPlot(SeuInt_2, features  = c("percent.mt"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
QC_int4 <- VlnPlot(SeuInt_2, features  = c("percent.rb"), ncol = 2, pt.size = 0, split.by = "orig.ident", split.plot = TRUE) + xlab(NULL) + theme(plot.title = element_text(size = 6, face = "bold"), axis.text = element_text(size = 5), legend.position = "none", plot.margin = unit(rep(0, 4), "npc"))
plot_grid(QC_int1, QC_int2, legend, QC_int3, QC_int4, nrow = 2, ncol = 3, rel_widths = c(1.2,1.2, 0.1), scale = 1)
```

```{r QCintegratedDS_separateRep2, out.width= '100%', fig.asp= 0.6}
QC_plots(subset(SeuInt_2, subset = orig.ident == "`3rd`"), "3rd DS", colors_n)

QC_plots(subset(SeuInt_2, subset = orig.ident == "`4th`"), "4th DS", colors_n)
```


## Make differential expression of neurons and then a dot plot to see if there are similar markers to the ones in Caramia 2023 paper.

```{r}
SeuInt_2_Ctrl <- subset(SeuInt_2, subset = Treat == "Ctrl")
DefaultAssay(SeuInt_2_Ctrl) <- "RNA"
SeuInt_2_n_Ctrl <- subset(SeuInt_2_Ctrl, subset = seurat_clusters != "3" & seurat_clusters != "5" & seurat_clusters != "6" & seurat_clusters != "7")
SeuInt_2_n_Ctrl_markers <- FindAllMarkers(SeuInt_2_n_Ctrl, test.use = "MAST", latent.vars = c("orig.ident", "Age"), only.pos =  T)
tops_2and3 <- DotPlotRemake(SeuInt_2_n_Ctrl, SeuInt_2_n_Ctrl_markers)
```

The clusters are less defined by the markers, but this could be a cause of having more clusters.

-   cluster 19 (ex cluster 15) remains the same. Marked by monaminergic and noradrenergic transports or noradrenaline biosinthetic enzymes.

-   cluster 23 (ex cluster 18) remains the same. Marked by Nos1, nitric oxide synthase, Fgfr1, fibroblast growth factor, and Slc6a5, a choline transpoter, **maybe acetylcholinergic cells**.

-   cluster 8 (ex cluster 5) remains the same. Marked by Arpp21, "The encoded protein is enriched in the caudate nucleus and cerebellar cortex. A similar protein in mouse may be involved in regulating the effects of dopamine in the basal ganglia" and Tcf7l2, " plays a key role in the Wnt signaling pathway" and Nrgn, "postsynaptic protein kinase substrate that binds calmodulin in the absence of calcium" **maybe a new cluster**.

-   cluster 13 (ex cluster 10) remains the same. Marked by mt-Co1, mt-Atp6 and Zic1, development. Is this a product of the technique, resembling many mRNAs in the sequencing "soup"?

-   cluster 24 (ex cluster 19) remains the same. Marked by EGFP-KASH, 1600010M07Rik, and Gm41333 if I include 3rd top gene, which probably means that this cluster **is a product of viral infection itself** and not something biologically meaninful

-   cluster 14 (ex cluster 13) remains the same. Marked by vomeronasal genes.

-   cluster 16 (ex cluster 7) slightly change. Marked by Dach1, development, Tcf4, ERK signaling, and Cacna2d3, calcium voltage-gated channel. 

-   cluster 18 (ex cluster 14) slightly change. Marked by 2600014E21Rik, lncRNA and Sst, somatostatin.

-   cluster 11 (ex cluster 9) remains the same. Marked by Sema5a, axonogenesis, Adarb2, RNA editing, and Ebf2, development.

-   cluster 2 (ex cluster 0). Marked, sigthly, by Bnc2, involved in differentiation, and Hs3st4, an heparan sulfate transferase.

-   cluster 1 (ex cluster 0). Not differentiated at all.

-   cluster 0 (ex cluster 0). Not differentiated at all.

-   cluster 9 (ex cluster 6) remains the same. Marked by Pbx1, development in general and associated to dopaminergic specification,  and Satb1, development and ssociated wuth adult subventricular zona development and Nfib.

-   cluster 15 (ex cluster 7) silghtly change. Marked by Cdh8, "putatively involved in synaptic adhesion, axon outgrowth and guidance", Ar, androgen receptor, and Gm49678

-   cluster 4 (ex cluster 0). Marked by Esrrg, an extrogen related receptor, Syt2, synaptotagmin2, and Kcna1, potassium volateg-gated channel.

-   cluster 10 (ex cluster 11) slightly change. Marked slightly by Nr4a2, hormone related to Parkinson Disease and essential for essential for development and mantainance of mdDA neurons, and C1ql4, maybe involved in synapse or differentiation.

-   cluster 12 (ex cluster 8) remains the same. Marked by Pde4d, a second menssager, slightly by Lmx1a, associated with ParkinsonÅ› Disease and important in development of dopamine producing neurons during embryogenesis and Chst9, a sulfotransferrase.

-   cluster 17 (ex cluster 12) remains the same. Marked by Zfp804b, TF enable to bind ion and act in the nucleus, Epha3, ERK signaling, Ghr, PI3K-Akt signaling.

## Make differential expression of neurons with an equilibrated DS (less cells for cluster 0, 1, 2 and 4) and then a dot plot to see if there are similar markers to the ones in Caramia 2023 paper.

```{r}
table(SeuInt_2_n_Ctrl$seurat_clusters, SeuInt_2_n_Ctrl$Age)
cells_df <- data.frame(seurat_clusters = SeuInt_2_n_Ctrl$seurat_clusters, cells = Cells(SeuInt_2_n_Ctrl))
cluster_0_cells <- filter(cells_df, seurat_clusters == "0")$cells
cluster_1_cells <- filter(cells_df, seurat_clusters == "1")$cells
cluster_2_cells <- filter(cells_df, seurat_clusters == "2")$cells
cluster_4_cells <- filter(cells_df, seurat_clusters == "4")$cells
other_cells <- filter(cells_df, seurat_clusters != "0" & seurat_clusters != "1" & seurat_clusters != "2" & seurat_clusters != "4")$cells
SeuInt_2_n_Ctrl_small <- subset(SeuInt_2_n_Ctrl, cells = c(other_cells, sample(cluster_0_cells, 100), sample(cluster_1_cells, 100), sample(cluster_2_cells, 100), sample(cluster_4_cells, 100)))

DimPlot(SeuInt_2_n_Ctrl_small, group.by = "seurat_clusters", order = T, label = T, cols = colors_n) +  theme_void(base_size = 10) + NoLegend() 
table(SeuInt_2_n_Ctrl_small$seurat_clusters, SeuInt_2_n_Ctrl_small$Age)

SeuInt_2_n_Ctrl_markers_small <- FindAllMarkers(SeuInt_2_n_Ctrl_small, test.use = "MAST", latent.vars = c("orig.ident", "Age"), only.pos =  T)
tops_2and3_small <- DotPlotRemake(SeuInt_2_n_Ctrl_small, SeuInt_2_n_Ctrl_markers_small)
```

Now the clusters that sligthly change:

-   cluster 18: more simmilar to ex cluster 14 given it is marked by 2600014E21Rik and Zfpm2

-   cluster 11: Ebf2 change for Alk, a tyrosine kinase

-   cluster 15: Ar change for Gm38505

-   cluster 14: Gm32467 change for Gm11309

-   cluster 2: Kcna1 change for 6330411D24Rik

-   cluster 12: Chst9 change for Zfhx3, a TF.

Given this, I would not consider this equalized DS and take with a grain of salt the top genes that change between whole DS and equalized DS

## For each month

For 1mo
```{r}
tops_2and3_small <- DotPlotRemake2(subset(SeuInt_2_n_Ctrl, subset = Age == "1mo"), SeuInt_2_n_Ctrl_markers)
```

For 2mo
```{r}
tops_2and3_small <- DotPlotRemake2(subset(SeuInt_2_n_Ctrl, subset = Age == "2mo"), SeuInt_2_n_Ctrl_markers)
```


## Some info about these markers

If I compare this with the data of Caramia 2023, I can see that all clusters resemble some already described cluster, by exception of:

-   cluster 8: **maybe a new cluster**

-   cluster 24:  **is a product of viral infection itself** and not something biologically meaninful. 

-   cluster 16: **cells under development**?

-   cluster 13: **is a product of viral infection itself** or **cells under development**?

-   cluster 23: **maybe acetylcholinergic cells**?

-   Cluster 14: only in the equalized DS express Gm32647, which was comparable to Caramia's data, comparison not trustable.

-   Cluster 16: Tcf4, characteristic of some clusters in Carmia's data.

The clusters that are already present in Caramia's lab are marked by:

-   cluster 9: by Nfib, Pbx1. (Carmia's cluster 11 or 22) **in development**

-   cluster 11: by Sema5a. (Carmia's cluster 7, 14 or 18) **in development**

-   cluster 12: by Zfhx3, but not trustable. (Carmia's cluster 17) **in development**

-   cluster 13: by Gm32647, but not trustable. (Caramia's cluster 17) **in olfatory system**?

-   cluster 14: by Zfpm2, but not trustable. (Caramia's cluster 2) **in development**

-   cluster 12: by Zfp804b. (Caramia's cluster 11, 15 or 18). **?**


## How top genes are seen

Define top genes as the ones that has a absolute log2FC stronger than 1

```{r}
SeuInt_2_n_Ctrl_markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) -> tops

DoHeatmapRemake(SeuInt_2_n_Ctrl, assesed_genes = unique(tops$gene), 2.5, colors_n, which_type = "Tops", gene_names = F)
```

Just addressing the top5 genes

```{r}
slice_head(tops, n = 5) %>%
    ungroup() -> top5

DoHeatmapRemake(SeuInt_2_n_Ctrl, assesed_genes = unique(top5$gene), 2.5, colors_n, which_type = "Top 5", gene_names = T)
```

Same but for the equalized DS, tops (but retrieved from whole DS)

```{r}
DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(tops$gene), 2.5, colors_n, which_type = "Tops", gene_names = F)
```

Same but for the equalized DS, top 5 (but retrieved from whole DS)

```{r}
DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(top5$gene), 2.5, colors_n, which_type = "Top 5", gene_names = T)
```

For the 1mo
```{r}
DoHeatmapRemakeAsSeurat(subset(SeuInt_2_n_Ctrl_small, subset = Age == "1mo"), markers = SeuInt_2_n_Ctrl_markers_small,  assesed_genes = unique(top5$gene), 2.5, colors_n, which_type = "Top 5", gene_names = T)
```

For the 2mo
```{r}
DoHeatmapRemakeAsSeurat(subset(SeuInt_2_n_Ctrl_small, subset = Age == "2mo"), markers = SeuInt_2_n_Ctrl_markers_small, assesed_genes = unique(top5$gene), 2.5, colors_n, which_type = "Top 5", gene_names = T)
```

#### Are these clusters differentially differentiated?

"Cell Fate" pathway genes

```{r}
pw_genes <- readRDS("Outputs/pw_genes_fate_diff.rds")
pw_genes <- do.call(rbind, pw_genes)
tops$gene[tops$gene %in% unique(pw_genes$SYMBOL)]
DoHeatmapRemake(SeuInt_2_n_Ctrl, assesed_genes = unique(pw_genes$SYMBOL), 2.5, colors_n, which_type = "Cell Fate", gene_names = T)
```

"Neuronal Development" pathway genes

```{r}
pw_genes <- readRDS("Outputs/pw_genes_develop_diff.rds")
pw_genes <- do.call(rbind, pw_genes)

DoHeatmapRemake(SeuInt_2_n_Ctrl, assesed_genes = unique(pw_genes$SYMBOL), 2.5, colors_n, which_type = "Neuronal Development", gene_names = F)
```


"Neuronal Development" pathway genes from tops

```{r}
DoHeatmapRemake(SeuInt_2_n_Ctrl, assesed_genes = unique(tops$gene[tops$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```

Ploted as Seurat

```{r}
DoHeatmapRemakeAsSeurat(SeuInt_2_n_Ctrl, markers = tops, assesed_genes = unique(tops$gene[tops$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```

% of DEG associated to development from total DEG

```{r}
length(tops$gene[tops$gene %in% unique(pw_genes$SYMBOL)])/length(tops$gene)
```

Just the 9% of DEGs are developmental genes.

And the dot plot of these genes is

```{r}
SeuInt_2_n_Ctrl_markers_reduced <- filter(SeuInt_2_n_Ctrl_markers, gene %in% unique(pw_genes$SYMBOL))
tops_2and3 <- DotPlotRemake(SeuInt_2_n_Ctrl, SeuInt_2_n_Ctrl_markers_reduced)
```

For 1mo
```{r}
tops_2and3_small <- DotPlotRemake(subset(SeuInt_2_n_Ctrl_small, subset = Age == "1mo"), SeuInt_2_n_Ctrl_markers_reduced)
```

For 2mo
```{r}
tops_2and3_small <- DotPlotRemake(subset(SeuInt_2_n_Ctrl_small, subset = Age == "2mo"), SeuInt_2_n_Ctrl_markers_reduced)
```


#### The same but for the equalized DS


"Neuronal Development" pathway genes from tops

```{r}
SeuInt_2_n_Ctrl_markers_small %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1, p_val_adj <= 0.05) -> tops_small

DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```

Ploted as Seurat

```{r}
DoHeatmapRemakeAsSeurat(SeuInt_2_n_Ctrl_small, markers = tops_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```

For the 1mo
```{r}
DoHeatmapRemakeAsSeurat(subset(SeuInt_2_n_Ctrl_small, subset = Age == "1mo"), markers = SeuInt_2_n_Ctrl_markers_small,  assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```

For the 2mo
```{r}
DoHeatmapRemakeAsSeurat(subset(SeuInt_2_n_Ctrl_small, subset = Age == "2mo"), markers = SeuInt_2_n_Ctrl_markers_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuronal Development", gene_names = T)
```


"Axonogenesis" pathway from tops_small

```{r}
pw_genes <- readRDS("Outputs/pw_genes_axonogenesis.rds")
pw_genes <- do.call(rbind, pw_genes)

DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Axonogenesis", gene_names = T)
```

Ploted as Seurat

```{r}
DoHeatmapRemakeAsSeurat(SeuInt_2_n_Ctrl_small, markers = tops_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Axonogenesis", gene_names = T)
```

"Neuron Projection Morphogenesis" pathway from tops_small

```{r}
pw_genes <- readRDS("Outputs/pw_genes_neu_project_morpho.rds")
pw_genes <- do.call(rbind, pw_genes)

DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuron Projection Morphogenesis", gene_names = T)
```

Ploted as Seurat

```{r}
DoHeatmapRemakeAsSeurat(SeuInt_2_n_Ctrl_small, markers = tops_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Neuron Projection Morphogenesis", gene_names = T)
```

"Cerebellar Development" pathway from tops_small

```{r}
pw_genes <- readRDS("Outputs/pw_genes_cerebellar_develop.rds")
pw_genes <- do.call(rbind, pw_genes)

DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Cerebellar Development", gene_names = T)
```

Ploted as Seurat

```{r}
DoHeatmapRemakeAsSeurat(SeuInt_2_n_Ctrl_small, markers = tops_small, assesed_genes = unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)]), 2.5, colors_n, which_type = "Tops in Cerebellar Development", gene_names = T)
```


"Gliogenesis, individual terms" pathway

```{r}
pw_genes <- readRDS("Outputs/pw_genes_gliogenesis_individual_terms.rds")
pw_genes <- do.call(rbind, pw_genes)
unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)])
DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(pw_genes$SYMBOL), 2.5, colors_n, which_type = "Gliogenesis, individual terms", gene_names = T)
```


"Noradrenergic Development, individual terms" pathway

```{r}
pw_genes <- readRDS("Outputs/pw_genes_noradrenergic_development.rds")
pw_genes <- do.call(rbind, pw_genes)
unique(tops_small$gene[tops_small$gene %in% unique(pw_genes$SYMBOL)])

DoHeatmapRemake(SeuInt_2_n_Ctrl_small, assesed_genes = unique(pw_genes$SYMBOL), 2.5, colors_n, which_type = "Noradrenergic Development, individual terms", gene_names = T)
```

## And to which pathways are associated the DEGs, in general?

```{r}
SeuInt_2_n_Ctrl_markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) -> tops_complete

all_genes_expressed <- rownames(SeuInt_2_n_Ctrl)[apply(GetAssayData(SeuInt_2_n_Ctrl, slot = "data"), 1, sum) > 0]

saveRDS(all_genes_expressed, paste0("Outputs/", date, "_QC_first_part_tidy_all_genes_expressed_LC.rds"))
saveRDS(tops_complete, paste0("Outputs/", date, "_QC_first_part_tidy_tops_DEGs_complete_LC.rds"))
saveRDS(tops, paste0("Outputs/", date, "_QC_first_part_tidy_tops_DEGs_LC.rds"))
saveRDS(colors_n, paste0("Outputs/", date, "_QC_first_part_tidy_cluster_colors_LC.rds"))
```

## UMAP for some DE genes which are also develmental genes:
```{r}
FeaturePlot(SeuInt_2_n_Ctrl, features = c("Nfib", "Sema5a", "Tcf7l2", "B2m", "Lmx1a", "Nr4a2", "Gal", "Ptprm", "Tgfbr2"))
```


Output to Make differential expression of neurons analysis
```{r}
saveRDS(SeuInt_1, paste0("Outputs/", date, "_SeuInt_3rd4th_fil_int_mayt1000lowt60000_lowtInfrb_19clusters.rds"))
saveRDS(SeuInt_2, paste0("Outputs/", date, "_SeuInt_3rd4th_fil_int_mayt1000lowt60000_lowtInfrb_24clusters.rds"))
```